### ISSUE TITLE
Public methods for add/remove slides

### ISSUE THREAD

**Issue by iosscripts** [iosscripts/iosslider#4]
Add public methods to add and remove slides.

**Comment by iosscripts** [comment]
I have added 3 new public methods. I will be adding them to the documentation once I feel they have been tested enough but for now I will just describe them here.

update: this function basically re-calibrates the slider manually. Sort of a catch all function for users who are manipulating parts of the slider/slides and need to recalculate the scrollbars, default css, etc.

``` js
$('.iosSlider').iosSlider('update');
```

addSlide: adds a new slide.

``` js
slideHTML = "<div class = 'slide'></div>":
slidePosition = 2;
$('.iosSlider').iosSlider('addSlide', slideHTML, slidePosition);
```

removeSlide: removes a slide from a given offset where. 

``` js
slidePosition = 2;
$('.iosSlider').iosSlider('removeSlide', slidePosition);
```

**Comment by iosscripts** [comment]
More testing is underway but I have pushed the changes to v1.0.12 and can be downloaded from the master branch now. Please let me know if this function lacks flexibility in any way and I will try to address it.

**Comment by dromonaco** [comment]
Thank you Marc. I will try the new methods out tomorrow.

-David Romo

On Jul 12, 2012, at 9:10 PM, Marc Whitbread
reply@reply.github.com
wrote:

**Comment by dromonaco** [comment]
I'm having 2 problems. 

The first is that the removeSlide method will not remove slides after there are only 3 remaining. I don't know if this has anything to do with it or not but our app shows 3 slides at a time.

The second problem I'm having is with the add method. I can't get it too work but I think that is because I am passing the HTML incorrectly. Normally to insert HTML on-the-fly I create jQuery objects and then either append or prepend them to existing DOM objects. For example:

``` javascript
   var slide = $("<div/>", {
      "class": "slide",
       "id": id
   });
```

Then I would add this to the DOM by calling:

``` javascript
$('#package_slider').prepend(slide);
```

Where package_slider is our div within the iosSlider div. I've replaced the prepend statement above with:

``` javascript
var html = slide.clone().wrap("<p>").parent().html();
$(".iosSlider").iosSlider('addSlide', html, 1);
```

For some reason this is not working for me. I'm going to continue investigating both problems.

**Comment by iosscripts** [comment]
To add, all you have to do is the following:

``` js
var slide = $("<div/>", {
    'class': 'item item1',
    'id': 'item1'
});
```

OR

``` js
var slide = "<div class = 'item item1' id = 'item1'></div>";
```

Then add it via the public method:

``` js
$('.iosSlider').iosSlider('addSlide', slide, 2);
```

The public method handles all the appending for you.

**Comment by iosscripts** [comment]
I will look at the removeSlide function not working below 3 slides.

**Comment by dromonaco** [comment]
For some reason the addSlide method doesn't work with our code. In my last comment I was using the jQuery HTML method because I thought that perhaps the addSlide method didn't work with a jQuery object but required HTML. I switched back to passing a jQuery object representing the slide in our call to the addSlide method but it is still not working. I also tried using a slidePosition of 2 instead of 1 just in case that was the problem. If you like I can post all for our code for the slide object. In the meantime I'm going to try to simplify our code for the slide object to be just one div.

**Comment by iosscripts** [comment]
If you don't mind posting the code that would be great. Take a look at the last benchmark in the _demos/_benchmarks/_general/index.html file. It has a working test case.

**Comment by dromonaco** [comment]
I was able to create this test case that fails for me using the google chrome or firefox browser on unbuntu 12.04.
testslider.css:

``` css

@charset "UTF-8";
/* slider container */
.iosSlider {
   /* required */
   position: relative;
   top: 0;
   left: 0;
   overflow: hidden;
   /* optional */
   width: 630px;
   height: 214px;
}
/* slider */
.iosSlider .slider {
   /* required */
   width: 100%;
   height: 100%;
}
/* slide */
.iosSlider .slider .slide {
   /* required */
   float: left;

   /* optional */
   width: 200px;
   height: 214px;
}
```

testslider.html

``` html

<html>
<head>
<meta name="format-detection" content="telephone=no" />
<meta name="format-detection" content="email=no" />
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
<meta http-equiv="X-UA-Compatible" content="IE=edge" />
<meta name = "viewport" content = "width=device-width, minimum-scale=1, maximum-scale=1">
<meta name = "apple-mobile-web-app-capable" content = "yes" />
<link href="testslider.css" rel="stylesheet" type="text/css" />
<script src="js/jquery-1.4.3.min.js" type="text/javascript"></script>
<script type = 'text/javascript' src = 'js/jquery.iosslider.min.js'></script>
<script type="text/javascript">
$(document).ready(function(){
   /* basic - default settings */
   $('.iosSlider').iosSlider({
      desktopClickDrag: true
   });
   $("#add_image").click(function(event) {
      var slide = $("<div />", {
         "class": "slide"
         }).append($("<img />", {
               "src": "images/florida/button-home-workflow.png"
               }));
      $(".iosSlider").iosSlider("addSlide", slide, 1);
   });
});
</script>
</head>
<body>
   <!-- slider container -->
   <div class = 'iosSlider'>
      <!-- slider -->
      <div class = 'slider'>
         <!-- slides -->
         <div class = 'slide'><img src="images/florida/button-home-deployment.png" /></div>
         <div class = 'slide'><img src="images/florida/button-home-explorer.png" /></div>
         <div class = 'slide'><img src="images/florida/button-home-history.png" /></div>
         <div class = 'slide'><img src="images/florida/button-home-workflow.png" /></div>
      </div>
   </div>
   <div id="add_image"><p>Click this to add an image</p></div>
</body>
</html>
```

I will try out your last benchmark later today when I get a chance.

**Comment by iosscripts** [comment]
Committed a possible fix to v1.0.15 on master.

**Comment by dromonaco** [comment]
I tested out the fix in v1.0.15. Add is now adding slides most of the time. Only the following issues remain:
- Add does not work if all the slides are visible at the same time. For example, if the view-able area is large enough to display 3 slides at once then add will not work if there are a total of 0 - 3 slides. This problem is reproducible with the code I posted above. The problem is also reproducible with the add/remove test case in _demos/_benchmarks/_general/index.html, by clicking remove until only 1 slide remains.
- Similarly, remove also does not work if all the slides are visible.
- The last issue I have is that addSlide(slide, 1), adds the new slide off-screen. Although, calling goToSlide(1) just after adding the slide should resolve this.

**Comment by marcwhitbread** [comment]
Thanks for the help with the testing. I will look into items 1 and 2. 

Regarding your third point; eg. if you are on slide 1 and you add a slide at position 1, you are automatically moved to slide 2 (previously slide 1). This is to make sure the user can add slides without being bumped around. Do you think the user should be automatically moved by default? or should it stay at slide 1?

**Comment by dromonaco** [comment]
Unfortunately I think both moving and not moving are the correct answers, depending on the context. In our case the add happens because a user clicks an add button so the user expects to see the slide they just added, but if slides are being added by some background process I would not expect to jump to the newly added slide.

**Comment by ColColonCleaner** [comment]
Just noticed this issue, and there is a connection with mine. The undefined error is happening when i have 3 slides on screen. I had no idea 3 was a magic number in your code somehow so i didn't test against it. I will add more slides and try my mod add/remove methods again.

Am i wrong to connect these dots in theory?

**Comment by iosscripts** [comment]
Anything is possible! Lol. Maybe thats what it is :S

Sent from Samsung Mobile

-------- Original message --------
Subject: Re: [iosSlider] Public methods for add/remove slides (#4)

**Comment by ColColonCleaner** [comment]
Confirmed that is the problem. Tested with 6 slides, removing 2 then adding 1 back works fine. Testing with 5 slides, removing 2, then trying to add 1 back fails. So yeah 3 is a magic number in your code somewhere.

Referencing #65

**Comment by iosscripts** [comment]
Fix applied to v1.0.36. Confirmed.
